### PR TITLE
Add additional operations to TorchTensor

### PR DESCRIPTION
**Pull request by dsyme**
This fills in some missing operations on TorchTensor.

@interesaaat Let me know what you think

### CODE REVIEW

File: src/TorchSharp/NN/ReLu.cs [hunk 1]
```diff
@@ -16,12 +16,9 @@ internal ReLU(bool inPlace = false) : base()
             _inPlace = inPlace;
         }
 
-        [DllImport("LibTorchSharp")]
-        private static extern IntPtr THSNN_reluApply(IntPtr tensor, bool inPlace);
```
Comment (dsyme) [inline src/TorchSharp/NN/ReLu.cs:20]: This signature doesn't match the signature in the native DLL (which has no inPlace argument)

File: src/TorchSharp/Tensor/TorchTensor.cs [hunk 2]
```diff
@@ -382,11 +407,11 @@ public TorchTensor Transpose(long dimension1, long dimension2)
         }
 
         [DllImport("LibTorchSharp")]
-        private static extern void THSTensor_transpose_(IntPtr src, long dim1, long dim2);
+        private static extern IntPtr THSTensor_transpose_(IntPtr src, long dim1, long dim2);
```
Comment (dsyme) [inline src/TorchSharp/Tensor/TorchTensor.cs:410]: The underlying native entrypoint returns an IntPtr, so putting that here too.

**Comment by interesaaat** [comment]
Thanks @dsyme for the PR. Looks good to me, but for some reason it is failing a test for `SplitWithSizesTest` on the mac build. Do you think you can look into it? I will merge afterwards.

**Comment by dsyme** [comment]
Just recording this for posterity: I had a bug when passing an optional Tensor to the C++ API.  

The Torch C++ methods often use optional arguments, but we generally need to pass them all through from C# and make them optional in C#, filling in the right default values.

When a Tensor is optional however you can't pass null into the C++ API, you have to pass a pointer to an empty tensor.

Compare
```
Tensor THSTensor_conv_transpose1d(
    const Tensor input, const Tensor weight, const Tensor bias,
    const int64_t* stride, const int strideLength,
    const int64_t* padding, const int paddingLength,
    const int64_t* outputPadding, const int outputPaddingLength,
    const int64_t* dilation, const int dilationLength,
    int64_t groups)
{
    auto res =
        torch::conv_transpose1d(*input, *weight, (bias ? *bias : at::Tensor()),
            at::IntList(stride, strideLength),
            at::IntList(padding, paddingLength),
            at::IntList(outputPadding, outputPaddingLength),
            groups,
            at::IntList(dilation, dilationLength));
    return new torch::Tensor(res);
}
```
and
```
Tensor THSTensor_conv_transpose1d(
    const Tensor input, const Tensor weight, const Tensor bias,
    const int64_t* stride, const int strideLength,
    const int64_t* padding, const int paddingLength,
    const int64_t* outputPadding, const int outputPaddingLength,
    const int64_t* dilation, const int dilationLength,
    int64_t groups)
{
    auto res =
        torch::conv_transpose1d(*input, *weight, *bias,
            at::IntList(stride, strideLength),
            at::IntList(padding, paddingLength),
            at::IntList(outputPadding, outputPaddingLength),
            groups,
            at::IntList(dilation, dilationLength));
    return new torch::Tensor(res);
}
```
Note the bias is an optional Tensor.  The first formulation is correct.

**Comment by dsyme** [comment]
> Thanks @dsyme for the PR. Looks good to me, but for some reason it is failing a test for SplitWithSizesTest on the mac build. Do you think you can look into it? I will merge afterwards.

I'm looking into this - perhaps it is to do with the memory leak bug.

**Comment by interesaaat** [comment]
Was this meant to be closed or merged?

**Comment by dsyme** [comment]
I have updated this with 

* [x] adding exception handling for most TorchTensor tensor operations 
* [x] testing for convolutions
* [x] maxpool and conv_transpose functions
* [x] SByteTensor

**Comment by dsyme** [comment]
> Was this meant to be closed or merged?

This was a mistake, GitHub automatically closed it when I accidentally pushed the "ops1" branch to this repo instead of my personal one

**Comment by dsyme** [comment]
@interesaaat This is now ready - it appears the Mac bug was due to the memory corruptions when pinned objects got moved around by the GC

**Pull request merged**